### PR TITLE
ENH: to_datetime support iso week year (16607) Updated

### LINKED ISSUE #16607: to_datetime should support ISO week year
`to_datetime` does not currently seem to support `ISO week year` like `strptime` does:

```
In [38]: datetime.date(2016, 1, 1).strftime('%G-%V')
Out[38]: '2015-53'

In [39]: datetime.datetime.strptime(datetime.date(2016, 1, 1).strftime('%G-%V')+'-1', '%G-%V-%u')
Out[39]: datetime.datetime(2015, 12, 28, 0, 0)

In [41]: pd.to_datetime(datetime.date(2016, 1, 1).strftime('%G-%V')+'-1', format='%G-%V-%u')
        ---------------------------------------------------------------------------
        TypeError                                 Traceback (most recent call last)
        /Users/Robin/.pyenv/versions/3.6.1/lib/python3.6/site-packages/pandas/core/tools/datetimes.py in _convert_listlike(arg, box, format, name, tz)
            443             try:
        --> 444                 values, tz = tslib.datetime_to_datetime64(arg)
            445                 return DatetimeIndex._simple_new(values, name=name, tz=tz)

        pandas/_libs/tslib.pyx in pandas._libs.tslib.datetime_to_datetime64 (pandas/_libs/tslib.c:33275)()

        TypeError: Unrecognized value type: <class 'str'>

        During handling of the above exception, another exception occurred:

        ValueError                                Traceback (most recent call last)
        <ipython-input-41-7ce30c959690> in <module>()
        ----> 1 pd.to_datetime(datetime.date(2016, 1, 1).strftime('%G-%V')+'-1', format='%G-%V-%u')

        /Users/Robin/.pyenv/versions/3.6.1/lib/python3.6/site-packages/pandas/core/tools/datetimes.py in to_datetime(arg, errors, dayfirst, yearfirst, utc, box, format, exact, unit, infer_datetime_format, origin)
            516         result = _convert_listlike(arg, box, format)
            517     else:
        --> 518         result = _convert_listlike(np.array([arg]), box, format)[0]
            519 
            520     return result

        /Users/Robin/.pyenv/versions/3.6.1/lib/python3.6/site-packages/pandas/core/tools/datetimes.py in _convert_listlike(arg, box, format, name, tz)
            445                 return DatetimeIndex._simple_new(values, name=name, tz=tz)
            446             except (ValueError, TypeError):
        --> 447                 raise e
            448 
            449     if arg is None:

        /Users/Robin/.pyenv/versions/3.6.1/lib/python3.6/site-packages/pandas/core/tools/datetimes.py in _convert_listlike(arg, box, format, name, tz)
            412                     try:
            413                         result = tslib.array_strptime(arg, format, exact=exact,
        --> 414                                                       errors=errors)
            415                     except tslib.OutOfBoundsDatetime:
            416                         if errors == 'raise':

        pandas/_libs/tslib.pyx in pandas._libs.tslib.array_strptime (pandas/_libs/tslib.c:63124)()

        pandas/_libs/tslib.pyx in pandas._libs.tslib.array_strptime (pandas/_libs/tslib.c:63003)()

        ValueError: 'G' is a bad directive in format '%G-%V-%u'

```

<details>
INSTALLED VERSIONS
------------------
commit: None

pandas: 0.20.1
pytest: 3.1.0
pip: 9.0.1
setuptools: 28.8.0
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: None
IPython: 6.0.0
sphinx: None
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: 3.4.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.999999999
sqlalchemy: 1.1.10
pymysql: None
psycopg2: 2.7.1 (dt dec pq3 ext lo64)
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by RjLi13**
- [x] closes #16607 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I found the issue stagnant, but since the fix was already in place, I manually took the code @rosygupta made and applied it onto latest master. Rebase wasn't an option that I found would work since the file in question has been split into multiple. Let me know what else needs to be updated.

Question: Given the code was created from a reference of cpython https://github.com/python/cpython/blob/6f0eb93183519024cb360162bdd81b9faec97ba6/Lib/_strptime.py#L321 
 I'm wondering why we can't use cpython's strptime directly and have to make our own implementation?

### CODE REVIEW

**Comment by sdrees** [comment]
Failures in travisci build for python 3.5? 
https://travis-ci.org/pandas-dev/pandas/jobs/481950579 maybe this reduces in the project codecpv down to zero? Hth

File: doc/source/whatsnew/v0.24.0.rst [hunk 1]
```diff
@@ -362,6 +362,7 @@ See the :ref:`Advanced documentation on renaming<advanced.index_names>` for more
 Other Enhancements
 ^^^^^^^^^^^^^^^^^^
 
+- Added support for %G strptime directive in parsing datetimes
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.24.0.rst]: * Reference the issue number
* I think this description is a little too technical for our general end user.  Can we describe the enhancement without referencing the directive?

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 2]
```diff
@@ -247,6 +247,53 @@ def test_to_datetime_parse_timezone_keeps_name(self):
 
 
 class TestToDatetime(object):
+    @pytest.mark.parametrize("s, _format, dt", [
+        ['2015-1-1', '%G-%V-%u', datetime(2014, 12, 29, 0, 0)],
+        ['2015-1-4', '%G-%V-%u', datetime(2015, 1, 1, 0, 0)],
+        ['2015-1-7', '%G-%V-%u', datetime(2015, 1, 4, 0, 0)]
+    ])
+    def test_to_datetime_iso_week_year_format(self, s, _format, dt):
+        assert to_datetime(s, format=_format) == dt
```
Comment (gfyoung) [inline pandas/tests/indexes/datetimes/test_tools.py]: For all of your newly added tests, reference as a comment the issue number below the function definition.

Comment (RjLi13) [reply pandas/tests/indexes/datetimes/test_tools.py]: Done

File: doc/source/whatsnew/v0.24.0.rst (hunk 1 above)
Comment (RjLi13) [reply doc/source/whatsnew/v0.24.0.rst]: - Done
- Did change it into 'Added support for ISO week year format when parsing datetimes using `to_datetime`' if that works better

Comment (gfyoung) [reply doc/source/whatsnew/v0.24.0.rst]: Much better.  Thanks!

**Comment by RjLi13** [comment]
Hey so I am currently stuck on making the checks pass. From what I can understand from the CI checks, it seems on Linux builds the `test_ValueError_iso_week_year` is returning a different error message than expected. Somehow the error message is being caught in https://github.com/pandas-dev/pandas/blob/e99049a820bf6a914513f715ce7ed0b01224d826/pandas/_libs/tslibs/strptime.pyx#L157 instead of https://github.com/pandas-dev/pandas/blob/e99049a820bf6a914513f715ce7ed0b01224d826/pandas/_libs/tslibs/strptime.pyx#L282 and I'm unclear why this happens when a year ago the changes did pass that time's tests.

**Comment by gfyoung** [comment]
> a year ago the changes did pass that time's tests.

@RjLi13 : Our codebase changes relatively quickly, so a year is essentially an eternity in our world.  A lot could have changed since then.  Here's what I would do:

Take the code for your test setup (and the particular `pytest` case that it's failing) and run the code for that test in your Python console.  Then look at the stack-trace to see how the logic is being traced.  It could be that the logic isn't even getting to your error condition (because this `strptime` error happens earlier), or it is, but erroneously passing over it.

If you're having issues, feel free to share the stacktrace, and we can have a look.

**Comment by RjLi13** [comment]
@gfyoung  So I have a Mac running python3.7 and the Azure pipelines show that only Linux builds seems to fail. I don't know what OS travisCI runs on but I don't have this specific environment currently on my machine
```
Python: 3.5
JOB="3.6, locale" ENV_FILE="ci/deps/travis-36-locale.yaml" PATTERN="((not slow and not network) or (single and db))" LOCALE_OVERRIDE="zh_CN.UTF-8"
```
Because of this I can't reproduce the test failures. 

Could you give me some steps on how to recreate such an env or reproduce the failures?

**Comment by RjLi13** [comment]
Here's my console output on my dev environment. I followed the instructions to run my tests from the doc (they contain 'iso_week_year')

```
(pandas-dev) Ruijings-MacBook-Pro:pandas ruijing$ pytest pandas/tests/indexes/datetimes/test_tools.py -v -k iso_week_year
======================================= test session starts ========================================
platform darwin -- Python 3.7.2, pytest-4.1.1, py-1.7.0, pluggy-0.8.0 -- /Users/ruijing/anaconda3/envs/pandas-dev/bin/python
cachedir: .pytest_cache
hypothesis profile 'ci' -> deadline=500, timeout=unlimited, suppress_health_check=[HealthCheck.too_slow], database=DirectoryBasedExampleDatabase('/Users/ruijing/pandas/.hypothesis/examples')
rootdir: /Users/ruijing/pandas, inifile: setup.cfg
plugins: xdist-1.26.0, forked-0.2, cov-2.6.1, hypothesis-4.0.1
collected 379 items / 363 deselected                                                               

pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_to_datetime_iso_week_year_format[2015-1-1-%G-%V-%u-dt0] PASSED [  6%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_to_datetime_iso_week_year_format[2015-1-4-%G-%V-%u-dt1] PASSED [ 12%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_to_datetime_iso_week_year_format[2015-1-7-%G-%V-%u-dt2] PASSED [ 18%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_ValueError_iso_week_year[ISO week directive '%V' must be used with the ISO year directive '%G' and a weekday directive '%A', '%a', '%w', or '%u'.-1999 50-%Y %V] PASSED [ 25%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_ValueError_iso_week_year[ISO year directive '%G' must be used with the ISO week directive '%V' and a weekday directive '%A', '%a', '%w', or '%u'.-1999 51-%G %V] PASSED [ 31%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_ValueError_iso_week_year[ISO year directive '%G' must be used with the ISO week directive '%V' and a weekday directive '%A', '%a', '%w', or '%u'.-1999 Monday-%G %A] PASSED [ 37%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_ValueError_iso_week_year[ISO year directive '%G' must be used with the ISO week directive '%V' and a weekday directive '%A', '%a', '%w', or '%u'.-1999 Mon-%G %a] PASSED [ 43%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_ValueError_iso_week_year[ISO year directive '%G' must be used with the ISO week directive '%V' and a weekday directive '%A', '%a', '%w', or '%u'.-1999 6-%G %w] PASSED [ 50%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_ValueError_iso_week_year[ISO year directive '%G' must be used with the ISO week directive '%V' and a weekday directive '%A', '%a', '%w', or '%u'.-1999 6-%G %u] PASSED [ 56%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_ValueError_iso_week_year[ISO year directive '%G' must be used with the ISO week directive '%V' and a weekday directive '%A', '%a', '%w', or '%u'.-2051-%G] PASSED [ 62%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_ValueError_iso_week_year[Day of the year directive '%j' is not compatible with ISO year directive '%G'. Use '%Y' instead.-1999 51 6 256-%G %V %u %j] PASSED [ 68%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_ValueError_iso_week_year[ISO week directive '%V' is incompatible with the year directive '%Y'. Use the ISO year '%G' instead.-1999 51 Sunday-%Y %V %A] PASSED [ 75%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_ValueError_iso_week_year[ISO week directive '%V' is incompatible with the year directive '%Y'. Use the ISO year '%G' instead.-1999 51 Sun-%Y %V %a] PASSED [ 81%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_ValueError_iso_week_year[ISO week directive '%V' is incompatible with the year directive '%Y'. Use the ISO year '%G' instead.-1999 51 1-%Y %V %w] PASSED [ 87%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_ValueError_iso_week_year[ISO week directive '%V' is incompatible with the year directive '%Y'. Use the ISO year '%G' instead.-1999 51 1-%Y %V %u] PASSED [ 93%]
pandas/tests/indexes/datetimes/test_tools.py::TestToDatetime::test_ValueError_iso_week_year[ISO week directive '%V' must be used with the ISO year directive '%G' and a weekday directive '%A', '%a', '%w', or '%u'.-20-%V] PASSED [100%]

============================ 16 passed, 363 deselected in 0.17 seconds =============================
(pandas-dev) Ruijings-MacBook-Pro:pandas ruijing$ 
```

In addition, I ran `pytest pandas` and `pytest pandas/tests/indexes/datetimes/test_tools.py -v` with no failures reported. Please advise next steps.

**Comment by gfyoung** [comment]
@RjLi13 : Cool!  Thanks for doing that.  Alright, so what I'm observing is that all of the relevant failures occurred on Python 3.x environments with a locale override, which are:

* `zh_CN.UTF-8`
* `it_IT.UTF-8`

Do you mind [overriding](https://github.com/pandas-dev/pandas/blob/00f5abab2150a4e9e1798d07bafc40e734941611/ci/run_tests.sh#L15-L18) your locale first, double check that the override works, and then run those same test commands again?

File: pandas/_libs/tslibs/strptime.pyx [hunk 3]
```diff
@@ -265,13 +270,44 @@ def array_strptime(object[:] values, object fmt,
...
 
         # If we know the wk of the year and what day of that wk, we can figure
         # out the Julian day of the year.
-        if julian == -1 and week_of_year != -1 and weekday != -1:
-            week_starts_Mon = True if week_of_year_start == 0 else False
-            julian = _calc_julian_from_U_or_W(year, week_of_year, weekday,
-                                              week_starts_Mon)
+        if julian == -1 and weekday != -1:
+            if week_of_year != -1:
+                week_starts_Mon = True if week_of_year_start == 0 else False
```
Comment (mroeschke) [inline pandas/_libs/tslibs/strptime.pyx]: just `week_starts_Mon = week_of_year_start == 0`

Comment (RjLi13) [reply pandas/_libs/tslibs/strptime.pyx]: Nice catch!

**Comment by RjLi13** [comment]
@gfyoung  can you explain how to override the locale? I'm not quite understanding what you linked.

**Comment by gfyoung** [comment]
> But running pandas.get_option("display.encoding") still gives me 'utf-8'

Hmmm...I see...since it is successfully changed after you check with the `locale` command, can you try running tests anyway and see what happens?

**Comment by RjLi13** [comment]
I did run with the new locale on the `iso_week_year` tests added. But they all pass, so I'm not quite sure what to do now.

**Comment by gfyoung** [comment]
> so I'm not quite sure what to do now.

Hmm...that is very weird.  Let's see if updating your branch might do the trick then.  That would entail:

* Addressing all unresolved comments (there's at least one)
* Rebasing your changes on `master`
* Moving your `whatsnew` entry to `0.25.0`

Let's make all of those changes and re-trigger CI and see what happens.

**Comment by RjLi13** [comment]
Before that, I ran the full pytest pandas suite using just `pytest pandas` after overriding the locale to `zh_CN.UTF-8`. Interestingly, I do see 6 errors. The stacktrace is rather long and I see tests other than mine seem to have failed. It doesn't seem like the stacktrace that the CI test shows. This is part of the stacktrace of the failure that happened on my test.

```
============================================= FAILURES =============================================
 TestToDatetime.test_ValueError_iso_week_year[ISO year directive '%G' must be used with the ISO week directive '%V' and a weekday directive '%A', '%a', '%w', or '%u'.-1999 Monday-%G %A] 

arg = array(['1999 Monday'], dtype=object), box = True, format = '%G %A', name = None, tz = None
unit = None, errors = 'raise', infer_datetime_format = False, dayfirst = False, yearfirst = False
exact = True

    def _convert_listlike_datetimes(arg, box, format, name=None, tz=None,
                                    unit=None, errors=None,
                                    infer_datetime_format=None, dayfirst=None,
                                    yearfirst=None, exact=None):
        """
        Helper function for to_datetime. Performs the conversions of 1D listlike
        of dates
    
        Parameters
        ----------
        arg : list, tuple, ndarray, Series, Index
            date to be parced
        box : boolean
            True boxes result as an Index-like, False returns an ndarray
        name : object
            None or string for the Index name
        tz : object
            None or 'utc'
        unit : string
            None or string of the frequency of the passed data
        errors : string
            error handing behaviors from to_datetime, 'raise', 'coerce', 'ignore'
        infer_datetime_format : boolean
            inferring format behavior from to_datetime
        dayfirst : boolean
            dayfirst parsing behavior from to_datetime
        yearfirst : boolean
            yearfirst parsing behavior from to_datetime
        exact : boolean
            exact format matching behavior from to_datetime
    
        Returns
        -------
        ndarray of parsed dates
            Returns:
    
            - Index-like if box=True
            - ndarray of Timestamps if box=False
        """
        from pandas import DatetimeIndex
        from pandas.core.arrays import DatetimeArray
        from pandas.core.arrays.datetimes import (
            maybe_convert_dtype, objects_to_datetime64ns)
    
        if isinstance(arg, (list, tuple)):
            arg = np.array(arg, dtype='O')
    
        # these are shortcutable
        if is_datetime64tz_dtype(arg):
            if not isinstance(arg, (DatetimeArray, DatetimeIndex)):
                return DatetimeIndex(arg, tz=tz, name=name)
            if tz == 'utc':
                arg = arg.tz_convert(None).tz_localize(tz)
            return arg
    
        elif is_datetime64_ns_dtype(arg):
            if box and not isinstance(arg, (DatetimeArray, DatetimeIndex)):
                try:
                    return DatetimeIndex(arg, tz=tz, name=name)
                except ValueError:
                    pass
    
            return arg
    
        elif unit is not None:
            if format is not None:
                raise ValueError("cannot specify both format and unit")
            arg = getattr(arg, 'values', arg)
            result = tslib.array_with_unit_to_datetime(arg, unit,
                                                       errors=errors)
            if box:
                if errors == 'ignore':
                    from pandas import Index
                    result = Index(result, name=name)
                    # GH 23758: We may still need to localize the result with tz
                    try:
                        return result.tz_localize(tz)
                    except AttributeError:
                        return result
    
                return DatetimeIndex(result, tz=tz, name=name)
            return result
        elif getattr(arg, 'ndim', 1) > 1:
            raise TypeError('arg must be a string, datetime, list, tuple, '
                            '1-d array, or Series')
    
        # warn if passing timedelta64, raise for PeriodDtype
        # NB: this must come after unit transformation
        orig_arg = arg
        arg, _ = maybe_convert_dtype(arg, copy=False)
    
        arg = ensure_object(arg)
        require_iso8601 = False
    
        if infer_datetime_format and format is None:
            format = _guess_datetime_format_for_array(arg, dayfirst=dayfirst)
    
        if format is not None:
            # There is a special fast-path for iso8601 formatted
            # datetime strings, so in those cases don't use the inferred
            # format because this path makes process slower in this
            # special case
            format_is_iso8601 = _format_is_iso(format)
            if format_is_iso8601:
                require_iso8601 = not infer_datetime_format
                format = None
    
        tz_parsed = None
        result = None
    
        if format is not None:
            try:
                # shortcut formatting here
                if format == '%Y%m%d':
                    try:
                        # pass orig_arg as float-dtype may have been converted to
                        # datetime64[ns]
                        orig_arg = ensure_object(orig_arg)
                        result = _attempt_YYYYMMDD(orig_arg, errors=errors)
                    except (ValueError, TypeError, tslibs.OutOfBoundsDatetime):
                        raise ValueError("cannot convert the input to "
                                         "'%Y%m%d' date format")
    
                # fallback
                if result is None:
                    try:
                        result, timezones = array_strptime(
                            arg, format, exact=exact, errors=errors)
                        if '%Z' in format or '%z' in format:
                            return _return_parsed_timezone_results(
                                result, timezones, box, tz, name)
                    except tslibs.OutOfBoundsDatetime:
                        if errors == 'raise':
                            raise
                        elif errors == 'coerce':
                            result = np.empty(arg.shape, dtype='M8[ns]')
                            iresult = result.view('i8')
                            iresult.fill(tslibs.iNaT)
                        else:
                            result = arg
                    except ValueError:
                        # if format was inferred, try falling back
                        # to array_to_datetime - terminate here
                        # for specified formats
                        if not infer_datetime_format:
                            if errors == 'raise':
                                raise
                            elif errors == 'coerce':
                                result = np.empty(arg.shape, dtype='M8[ns]')
                                iresult = result.view('i8')
                                iresult.fill(tslibs.iNaT)
                            else:
                                result = arg
            except ValueError as e:
                # Fallback to try to convert datetime objects if timezone-aware
                #  datetime objects are found without passing `utc=True`
                try:
>                   values, tz = conversion.datetime_to_datetime64(arg)

pandas/core/tools/datetimes.py:291: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   raise TypeError('Unrecognized value type: %s' % type(val))
E   TypeError: Unrecognized value type: <class 'str'>

pandas/_libs/tslibs/conversion.pyx:200: TypeError

During handling of the above exception, another exception occurred:

self = <pandas.tests.indexes.datetimes.test_tools.TestToDatetime object at 0x1c5894e8d0>
msg = "ISO year directive '%G' must be used with the ISO week directive '%V' and a weekday directive '%A', '%a', '%w', or '%u'."
s = '1999 Monday', _format = '%G %A'

    @pytest.mark.parametrize("msg, s, _format", [
        ["ISO week directive '%V' must be used with the ISO year directive "
         "'%G' and a weekday directive '%A', '%a', '%w', or '%u'.", "1999 50",
         "%Y %V"],
        ["ISO year directive '%G' must be used with the ISO week directive "
         "'%V' and a weekday directive '%A', '%a', '%w', or '%u'.", "1999 51",
         "%G %V"],
        ["ISO year directive '%G' must be used with the ISO week directive "
         "'%V' and a weekday directive '%A', '%a', '%w', or '%u'.", "1999 "
         "Monday", "%G %A"],
        ["ISO year directive '%G' must be used with the ISO week directive "
         "'%V' and a weekday directive '%A', '%a', '%w', or '%u'.", "1999 Mon",
         "%G %a"],
        ["ISO year directive '%G' must be used with the ISO week directive "
         "'%V' and a weekday directive '%A', '%a', '%w', or '%u'.", "1999 6",
         "%G %w"],
        ["ISO year directive '%G' must be used with the ISO week directive "
         "'%V' and a weekday directive '%A', '%a', '%w', or '%u'.", "1999 6",
         "%G %u"],
        ["ISO year directive '%G' must be used with the ISO week directive "
         "'%V' and a weekday directive '%A', '%a', '%w', or '%u'.", "2051",
         "%G"],
        ["Day of the year directive '%j' is not compatible with ISO year "
         "directive '%G'. Use '%Y' instead.", "1999 51 6 256", "%G %V %u %j"],
        ["ISO week directive '%V' is incompatible with the year directive "
         "'%Y'. Use the ISO year '%G' instead.", "1999 51 Sunday", "%Y %V %A"],
        ["ISO week directive '%V' is incompatible with the year directive "
         "'%Y'. Use the ISO year '%G' instead.", "1999 51 Sun", "%Y %V %a"],
        ["ISO week directive '%V' is incompatible with the year directive "
         "'%Y'. Use the ISO year '%G' instead.", "1999 51 1", "%Y %V %w"],
        ["ISO week directive '%V' is incompatible with the year directive "
         "'%Y'. Use the ISO year '%G' instead.", "1999 51 1", "%Y %V %u"],
        ["ISO week directive '%V' must be used with the ISO year directive "
         "'%G' and a weekday directive '%A', '%a', '%w', or '%u'.", "20", "%V"]
    ])
    def test_ValueError_iso_week_year(self, msg, s, _format):
        # See GH#16607
        with pytest.raises(ValueError, match=msg):
>           to_datetime(s, format=_format)

pandas/tests/indexes/datetimes/test_tools.py:297: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
```

I have a full copy of the failures stacktrace. It is 3000+ lines long. I'm not sure where to start debugging with it or should I even try posting it here.

**Comment by RjLi13** [comment]
I'll push the latest commit I have, and then let's think of what should be the next steps?

File: doc/source/whatsnew/v0.24.0.rst [hunk 4]
```diff
@@ -364,6 +364,7 @@ See the :ref:`Advanced documentation on renaming<advanced.index_names>` for more
 Other Enhancements
 ^^^^^^^^^^^^^^^^^^
 
+- Added support for ISO week year format when parsing datetimes using `to_datetime` (:issue:`16607`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.rst]: move to 0.25.0, use ``:meth:`to_datetime` ``

Comment (jreback) [reply doc/source/whatsnew/v0.24.0.rst]: be more explicit about which codes are added

File: pandas/_libs/tslibs/strptime.pyx [hunk 5]
```diff
@@ -511,14 +547,17 @@ class TimeRE(dict):
             # The " \d" part of the regex is to make %c from ANSI C work
             'd': r"(?P<d>3[0-1]|[1-2]\d|0[1-9]|[1-9]| [1-9])",
             'f': r"(?P<f>[0-9]{1,9})",
+            'G': r"(?P<G>\d\d\d\d)",
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: I am not sure we have a doc-refernce to the strptime codes anywhere @mroeschke do you know?

File: pandas/_libs/tslibs/strptime.pyx [hunk 6]
```diff
@@ -620,6 +659,25 @@ cdef _calc_julian_from_U_or_W(int year, int week_of_year,
         return 1 + days_to_week + day_of_week
 
 
+cdef _calc_julian_from_V(int iso_year, int iso_week, int iso_weekday):
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: can you add a more format doc-string (parameters / returns) sections

Comment (jreback) [reply pandas/_libs/tslibs/strptime.pyx]: if you can update the other _calc_julian one would be great as well

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 7]
```diff
@@ -247,6 +247,55 @@ def test_to_datetime_parse_timezone_keeps_name(self):
 
 
 class TestToDatetime(object):
+    @pytest.mark.parametrize("s, _format, dt", [
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_tools.py]: we should split up this file to segregate format parsing into a separate test file, cc @jbrockmendel, @mroeschke

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/strptime.pyx (hunk 5 above)
Comment (mroeschke) [reply pandas/_libs/tslibs/strptime.pyx]: Looks like the reference is here: https://github.com/pandas-dev/pandas/blob/fcc11feb1caf04061195d1b21c4952a4c31fa23a/pandas/_libs/tslibs/strptime.pyx#L361

Comment (jreback) [reply pandas/_libs/tslibs/strptime.pyx]: right so we have a doc reference anywhere though? eg in a doc string for Timestamp / to_datetime and / or in the docs proper?

Comment (mroeschke) [reply pandas/_libs/tslibs/strptime.pyx]: Yeah looks like the we link to the directives here: https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#providing-a-format-argument

**Comment by WillAyd** [comment]
@RjLi13 can you merge master and address comments?

File: doc/source/whatsnew/v0.24.0.rst (hunk 4 above)
Comment (WillAyd) [inline doc/source/whatsnew/v0.24.0.rst]: Move to 0.25 at this point

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by RjLi13** [comment]
Hey, sorry for the silence. Been busy these past few weeks. Will be pushing the rebased version and addressing changes this weekend.

**Comment by RjLi13** [comment]
Rebase gone wrong, not sure how to fix. I'm closing the PR and remaking a new one which I will reference here.

**Pull request closed without merging**